### DISCUSSION TITLE
Thoughts on extensibility

### DISCUSSION THREAD

**Discussion by weaversam8** [redimp/otterwiki#91] [Ideas]
OtterWiki has been a very solid option for wiki software that my team is interested in using. The small size and simplicity of the Python codebase has proven very easy to understand, which is powerful.

@redimp - what are your thoughts on extensibility in OtterWiki? Is a "plugin" API something that you feel fits in with your vision of the project? I think it could be a powerful way for users to extend the wiki without adding more complexity to the base project. A couple places we could add hooks for "plugins":

- After markdown is loaded from the Git repo, before it is rendered to HTML
- After HTML is rendered, before returning to the browser
- Pluggable storage engines
- Displaying information about a page in the sidebar (possibly from frontmatter)
- Adding action buttons to the dropdown menu on a page (or on the wiki as a whole)
- Pre-fill commit message with a plugin
- Markdown linting for style before commit (with Vale or other tools)

**Comment by redimp** [comment]
Hey @weaversam8,

A plugin api would be a welcome feature for An Otter Wiki. The point you're making is important to me: Keeping the core somewhat minimalistic. But I can understand that users want to add features, possibly a feature that a team team is desperately missing.

Finding places for hooks is not a problem. From adding Markdown features, to adding buttons/macros to the editor to changing the auth process are all things that I aleady thought about.

Can ypu suggest a good exmaple for python projectys with a solid plugin api? I would be happy to get some inspiration.

**Reply by weaversam8** [reply]
Sure! You might check out how [Pelican (a Python static site generator)][pelican] implements plugins, and you might consider using [pluggy (the plugin framework used by pytest)][pluggy] as a drop-in solution for building a plugin library.

[pelican]: https://docs.getpelican.com/en/latest/plugins.html
[pluggy]: https://pluggy.readthedocs.io/en/latest/

**Comment by weaversam8** [comment]
See #96 for a stab at solving this problem.

**Comment by redimp** [comment]
After #96 being merged we have the base to build plugins. Next questions/tasks that I could think of and would like to discuss:

- Allow plugins to integrate into preferences (make them configurable)
- Find all the places where hooks are useful
  - Find a way to let hooks extend templates, css, 
- Extend the docker image, so that it mounts a plugin directory, does a `pip install` on startup (respecting `PIP_INDEX_URL`)

Your list of ideas @weaversam8 is a good place to continue the work on this. Lets check what is necessary to add in hooks to the code, so that the ideas you mentioned can be integrated as plugins. 

After re-reading the list, I noticed that I have implemented "After HTML is rendered, before returning to the browser" with the `page_view_htmlcontent_postprocess()` hook.

**Comment by ReessKennedy** [comment]
Hey @weaversam8 - With very little code I added a Frontmatter/Page properties feature with very little code and submitted a pull request: https://github.com/redimp/otterwiki/pull/207

Not sure what @redimp will think but trying to do my part this world! :)

Continued thanks to @redimp for this cool project.
